File: analysis/ghidra_scripts/ReportUpdatePath.py
## ###
#  IP: GHIDRA
##
# Report update-path functions and references for a raw JJOS firmware image.
# @category: Analysis
# @runtime Jython

BASE = 0x09000000
from jarray import zeros

RANGES = [
    ("boot_update_block", BASE + 0x7be0, BASE + 0x8130, False),
    ("progress_block", BASE + 0x94a0, BASE + 0x94d0, False),
    ("progress_erase_exact", BASE + 0x94a0, BASE + 0x94a1, True),
    ("progress_write_exact", BASE + 0x94b8, BASE + 0x94b9, True),
]


def read_u32(memory, addr):
    data = zeros(4, 'b')
    read = memory.getBytes(addr, data)
    if read != 4:
        return -1
    return ((data[0] & 0xff) << 24) | ((data[1] & 0xff) << 16) | ((data[2] & 0xff) << 8) | (data[3] & 0xff)


def scan_for_refs(memory, start_value, end_value, exact_only):
    hits = []
    for block in memory.getBlocks():
        if not block.isInitialized():
            continue
        cur = block.getStart()
        end = block.getEnd().subtract(3)
        while cur.compareTo(end) <= 0 and not monitor.isCancelled():
            value = read_u32(memory, cur)
            if exact_only:
                match = value == start_value
            else:
                match = value >= start_value and value < end_value
            if match:
                hits.append(cur)
            cur = cur.add(2)
    return hits


def get_callers(func):
    out = []
    refs = currentProgram.getReferenceManager().getReferencesTo(func.getEntryPoint())
    while refs.hasNext():
        ref = refs.next()
        if ref.getReferenceType().isCall():
            out.append(ref.getFromAddress())
    out.sort(key=lambda x: x.getOffset())
    return out


def find_nearby_seed_function(ref_site):
    func = getFunctionContaining(ref_site)
    if func is not None:
        return func

    best = None
    best_delta = None
    funcs = currentProgram.getFunctionManager().getFunctions(True)
    while funcs.hasNext():
        candidate = funcs.next()
        if not candidate.getName().startswith("UPD_"):
            continue
        delta = ref_site.getOffset() - candidate.getEntryPoint().getOffset()
        if delta < 0 or delta > 0x40:
            continue
        if best is None or delta < best_delta:
            best = candidate
            best_delta = delta
    return best


def get_callees(func):
    out = []
    seen = set()
    inst = getInstructionAt(func.getEntryPoint())
    while inst is not None and func.getBody().contains(inst.getAddress()):
        if inst.getFlowType().isCall():
            for flow in inst.getFlows():
                callee = getFunctionAt(flow)
                if callee is not None:
                    item = "{}@{}".format(callee.getName(), callee.getEntryPoint())
                else:
                    item = "SUB@{}".format(flow)
                if item not in seen:
                    seen.add(item)
                    out.append(item)
        inst = inst.getNext()
    return out


def join_addresses(addrs):
    if addrs is None or len(addrs) == 0:
        return "-"
    return ", ".join([str(x) for x in addrs])


def main():
    out_path = None
    args = getScriptArgs()
    if len(args) > 0:
        out_path = args[0]

    memory = currentProgram.getMemory()
    listing = currentProgram.getListing()

    hits = {}
    for name, start_value, end_value, exact_only in RANGES:
        hits[name] = scan_for_refs(memory, start_value, end_value, exact_only)

    function_tags = {}
    function_sites = {}
    for tag in hits:
        for ref_site in hits[tag]:
            func = find_nearby_seed_function(ref_site)
            if func is None:
                continue
            function_tags.setdefault(func, set()).add(tag)
            function_sites.setdefault(func, []).append(ref_site)

    funcs = currentProgram.getFunctionManager().getFunctions(True)
    while funcs.hasNext():
        func = funcs.next()
        if func.getName().startswith("UPD_"):
            function_tags.setdefault(func, set())
            function_sites.setdefault(func, [])

    functions = sorted(function_tags.keys(), key=lambda f: f.getEntryPoint().getOffset())

    lines = []
    lines.append("# Ghidra Update Path Map")
    lines.append("")
    lines.append("- Program: `{}`".format(currentProgram.getName()))
    lines.append("- Processor: `68000:BE:32:Coldfire`")
    lines.append("- Image base: `0x09000000`")
    lines.append("")
    lines.append("## Reference Summary")
    lines.append("")
    for tag in hits:
        lines.append("- `{}` refs: `{}`".format(tag, len(hits[tag])))
    lines.append("")
    lines.append("## Seeded Function Map")
    lines.append("")

    for func in functions:
        sites = sorted(function_sites.get(func, []), key=lambda x: x.getOffset())
        callers = get_callers(func)
        callees = get_callees(func)
        lines.append("### `{}`".format(func.getName()))
        lines.append("")
        lines.append("- Entry: `{}`".format(func.getEntryPoint()))
        lines.append("- Tags: `{}`".format(", ".join(sorted(function_tags.get(func, [])))))
        lines.append("- Ref sites: `{}`".format(join_addresses(sites)))
        lines.append("- Callers: `{}`".format(join_addresses(callers)))
        lines.append("- Callees: `{}`".format(", ".join(callees) if len(callees) else "-"))
        lines.append("")

    lines.append("## Observations")
    lines.append("")
    lines.append("- Functions tagged with `progress_block` are the strongest current flash/write routine candidates.")
    lines.append("- Functions tagged only with `boot_update_block` are more likely verification, load/update UI, or error handling paths.")
    lines.append("- Zero exact refs to `progress_write_exact` means the write-progress string is probably reached indirectly or via a nearby literal pool entry.")
    lines.append("")

    if out_path is not None:
        out_file = open(out_path, "w")
        for line in lines:
            out_file.write(line + "\n")
        out_file.close()


main()

File: analysis/ghidra_scripts/ReportUpdatePath.java
//@category Analysis

import java.io.File;
import java.io.PrintWriter;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.LinkedHashMap;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

import ghidra.app.script.GhidraScript;
import ghidra.program.model.address.Address;
import ghidra.program.model.listing.Function;
import ghidra.program.model.listing.FunctionIterator;
import ghidra.program.model.listing.Instruction;
import ghidra.program.model.listing.Listing;
import ghidra.program.model.mem.Memory;
import ghidra.program.model.mem.MemoryBlock;
import ghidra.program.model.symbol.RefType;
import ghidra.program.model.symbol.Reference;
import ghidra.program.model.symbol.ReferenceIterator;

public class ReportUpdatePath extends GhidraScript {

	private static class RefRange {
		String name;
		long start;
		long end;
		boolean exactOnly;

		RefRange(String name, long start, long end, boolean exactOnly) {
			this.name = name;
			this.start = start;
			this.end = end;
			this.exactOnly = exactOnly;
		}
	}

	private static final long BASE = 0x09000000L;

	private static final RefRange[] RANGES = new RefRange[] {
		new RefRange("boot_update_block", BASE + 0x7be0, BASE + 0x8130, false),
		new RefRange("progress_block", BASE + 0x94a0, BASE + 0x94d0, false),
		new RefRange("progress_erase_exact", BASE + 0x94a0, BASE + 0x94a1, true),
		new RefRange("progress_write_exact", BASE + 0x94b8, BASE + 0x94b9, true),
	};

	@Override
	public void run() throws Exception {
		String outPath = getScriptArgs().length > 0 ? getScriptArgs()[0] : null;
		Memory memory = currentProgram.getMemory();
		Listing listing = currentProgram.getListing();

		Map<String, List<Address>> hits = new LinkedHashMap<>();
		for (RefRange range : RANGES) {
			hits.put(range.name, scanForReferences(memory, range));
		}

		Map<Function, Set<String>> functionTags = new LinkedHashMap<>();
		Map<Function, Set<Address>> functionRefSites = new LinkedHashMap<>();

		for (Map.Entry<String, List<Address>> entry : hits.entrySet()) {
			for (Address refSite : entry.getValue()) {
				Function function = listing.getFunctionContaining(refSite);
				if (function == null) {
					continue;
				}
				functionTags.computeIfAbsent(function, key -> new LinkedHashSet<>()).add(entry.getKey());
				functionRefSites.computeIfAbsent(function, key -> new LinkedHashSet<>()).add(refSite);
			}
		}

		List<Function> functions = new ArrayList<>(functionTags.keySet());
		Collections.sort(functions, Comparator.comparingLong(f -> f.getEntryPoint().getOffset()));

		List<String> lines = new ArrayList<>();
		lines.add("# Ghidra Update Path Map");
		lines.add("");
		lines.add("- Program: `" + currentProgram.getName() + "`");
		lines.add("- Processor: `68000:BE:32:Coldfire`");
		lines.add("- Image base: `0x09000000`");
		lines.add("");
		lines.add("## Reference Summary");
		lines.add("");
		for (Map.Entry<String, List<Address>> entry : hits.entrySet()) {
			lines.add("- `" + entry.getKey() + "` refs: `" + entry.getValue().size() + "`");
		}
		lines.add("");
		lines.add("## Seeded Function Map");
		lines.add("");

		for (Function function : functions) {
			lines.add("### `" + function.getName() + "`");
			lines.add("");
			lines.add("- Entry: `" + function.getEntryPoint() + "`");
			lines.add("- Tags: `" + join(functionTags.get(function)) + "`");
			lines.add("- Ref sites: `" + joinAddresses(functionRefSites.get(function)) + "`");
			lines.add("- Callers: `" + joinAddresses(getCallers(function)) + "`");
			lines.add("- Callees: `" + joinFunctionCalls(function) + "`");
			lines.add("");
		}

		lines.add("## Observations");
		lines.add("");
		lines.add("- Functions tagged with `progress_block` are the strongest current flash/write routine candidates.");
		lines.add("- Functions tagged only with `boot_update_block` are more likely to belong to verification, load/update UI, or error handling.");
		lines.add("- Missing `progress_write_exact` refs means the write-progress string may be reached through a nearby pointer or table, not a direct immediate.");
		lines.add("");

		if (outPath != null) {
			writeLines(outPath, lines);
		}
	}

	private List<Address> scanForReferences(Memory memory, RefRange range) throws Exception {
		List<Address> results = new ArrayList<>();
		for (MemoryBlock block : memory.getBlocks()) {
			if (!block.isInitialized()) {
				continue;
			}
			long endOffset = block.getEnd().getOffset() - 3;
			for (long off = block.getStart().getOffset(); off <= endOffset && !monitor.isCancelled(); off += 2) {
				Address addr = toAddr(off);
				long value = readUnsignedInt(memory, addr);
				boolean match = range.exactOnly ? value == range.start
						: (value >= range.start && value < range.end);
				if (match) {
					results.add(addr);
				}
			}
		}
		Collections.sort(results, Comparator.comparingLong(Address::getOffset));
		return results;
	}

	private long readUnsignedInt(Memory memory, Address address) throws Exception {
		byte[] bytes = new byte[4];
		int read = memory.getBytes(address, bytes);
		if (read != 4) {
			return -1;
		}
		return ((bytes[0] & 0xffL) << 24) | ((bytes[1] & 0xffL) << 16) | ((bytes[2] & 0xffL) << 8) |
			(bytes[3] & 0xffL);
	}

	private List<Address> getCallers(Function function) {
		List<Address> callers = new ArrayList<>();
		ReferenceIterator iterator =
			currentProgram.getReferenceManager().getReferencesTo(function.getEntryPoint());
		while (iterator.hasNext()) {
			Reference reference = iterator.next();
			if (reference.getReferenceType().isCall()) {
				callers.add(reference.getFromAddress());
			}
		}
		Collections.sort(callers, Comparator.comparingLong(Address::getOffset));
		return callers;
	}

	private String joinFunctionCalls(Function function) {
		Set<String> calls = new LinkedHashSet<>();
		Instruction instruction = getInstructionAt(function.getEntryPoint());
		while (instruction != null && function.getBody().contains(instruction.getAddress())) {
			if (instruction.getFlowType().isCall()) {
				Address[] flows = instruction.getFlows();
				for (Address flow : flows) {
					Function callee = getFunctionAt(flow);
					if (callee != null) {
						calls.add(callee.getName() + "@" + callee.getEntryPoint());
					}
					else {
						calls.add("SUB@" + flow);
					}
				}
			}
			instruction = instruction.getNext();
		}
		return calls.isEmpty() ? "-" : join(calls);
	}

	private String joinAddresses(Set<Address> values) {
		if (values == null || values.isEmpty()) {
			return "-";
		}
		List<Address> rows = new ArrayList<>(values);
		Collections.sort(rows, Comparator.comparingLong(Address::getOffset));
		List<String> rendered = new ArrayList<>();
		for (Address addr : rows) {
			rendered.add(addr.toString());
		}
		return String.join(", ", rendered);
	}

	private String join(List<Address> values) {
		if (values == null || values.isEmpty()) {
			return "-";
		}
		List<String> rendered = new ArrayList<>();
		for (Address addr : values) {
			rendered.add(addr.toString());
		}
		return String.join(", ", rendered);
	}

	private String join(Set<String> values) {
		if (values == null || values.isEmpty()) {
			return "-";
		}
		return String.join(", ", values);
	}

	private void writeLines(String outPath, List<String> lines) throws Exception {
		File outFile = new File(outPath);
		outFile.getParentFile().mkdirs();
		PrintWriter writer = new PrintWriter(outFile);
		try {
			for (String line : lines) {
				writer.println(line);
			}
		}
		finally {
			writer.close();
		}
	}
}
